auth/authenticationController.go

```diff
@@ -0,0 +1,173 @@
+		err := controller.server.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed {
+			log.Errorf("could not start web server: %s", err.Error())
+		}
+	}()
+	<-ctx.Done()
+}
+
+// shutdownHandler handles the shutdown action of the server.
+//noinspection GoUnusedParameter
+func (controller *realAuthenticationController) shutdownHandler(w http.ResponseWriter, r *http.Request, cancel func()) {
+	log.Info("Shutting web server down.")
+	http.Redirect(w, r, controller.configuration.AuthenticationEndpoint, http.StatusSeeOther)
+	cancel()
+}
+
+// authenticationHandler handles the authentication action of the server.
+func (controller *realAuthenticationController) authenticationHandler(w http.ResponseWriter, r *http.Request) {
+	id := r.URL.Query().Get("id")
+	secret := r.URL.Query().Get("secret")
+
+	if id != "" && secret != "" {
+		systemToken, err := controller.createSystemToken(id, secret)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			_, err = w.Write([]byte(err.Error()))
+			if err != nil {
+				log.Errorf(failedToWriteFormat, err.Error())
+			}
+			return
+		}
+
+		creds := core.Credentials{Username: systemToken.ID, Password: systemToken.Secret}
+		err = controller.store.Add(controller.configuration.CredentialsStore, &creds)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			_, err = w.Write([]byte(err.Error()))
+			if err != nil {
+				log.Errorf(failedToWriteFormat, err.Error())
+			}
+			return
+		}
+
+		log.Info("Successfully authenticated.")
+		http.Redirect(w, r, "/shutdown", http.StatusFound)
+	} else {
+		controller.redirectToAuthEndpoint(w, r)
+		return
+	}
+}
+
+func (controller *realAuthenticationController) redirectToAuthEndpoint(w http.ResponseWriter, r *http.Request) {
+	redirectURL := fmt.Sprintf("%s/instance-registration?ces_redirect_uri=http://%s", controller.configuration.AuthenticationEndpoint, r.Host)
+	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
+	http.Redirect(w, r, redirectURL, http.StatusSeeOther)
+}
+
+// createSystemToken creates a system token from setup tokens.
+func (controller *realAuthenticationController) createSystemToken(id string, secret string) (SystemToken, error) {
+	systemToken := SystemToken{}
+
+	instanceRegistrationUrl := controller.configuration.AuthenticationEndpoint + "/api/v1/instance-registrations/" + id
+
+	if controller.configuration.PreviousInstanceID != "" {
+		instanceRegistrationUrl += "?previousId=" + controller.configuration.PreviousInstanceID
+	}
+
+	client := controller.client
+
+	// post temporal token to get login credentials for dogu registry
+	var tokenSecret = []byte(`{"secret":"` + secret + `"}`)
+	request, err := http.NewRequest(http.MethodPost, instanceRegistrationUrl, bytes.NewBuffer(tokenSecret))
+	if err != nil {
+		return systemToken, fmt.Errorf("could not create backend request: %w", err)
+	}
+	request.Header.Set("Content-Type", "application/json")
+
+	resp, err := client.Do(request)
```

Comment:
🚫 **[checkstyle]** <bodyclose> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>response body must be closed